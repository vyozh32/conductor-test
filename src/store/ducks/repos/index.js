import axios from 'axios';
import endpoints from '../../../endpoints';

// Actions

export const GET_REPOS_ERROR = 'conductor/test/get-repos-data';
export const GET_REPOS_SUCCESS = 'conductor/test/get-repos-data';

// defaultState

const defaultReposState = {
	data: null,
	error: null
};

// Reducer
export default function(state = defaultReposState, action) {
	const { type, payload } = action;
	switch (type) {
		case GET_REPOS_SUCCESS: {
			return {
				...state,
				data: payload,
				error: null
			};
		}
		case GET_REPOS_ERROR: {
			return {
				...state,
				loading: false,
				loaded: true,
				error: payload
			};
		}
		default: {
			return state;
		}
	}
}

// Action Creators

const loadReposData = (type, payload) => {
	return {
		type: type,
		payload
	};
};

// side effects, only as applicable
// e.g. thunks, epics, etc

export const actions = {
	getReposData: () => {
		return (dispatch, getState) => {
			const user = getState().user.data.login;
			const url = endpoints.REPOS.replace('{user_id}', user);
			axios
				.get(url)
				.then((res) => {
					return {
						data: res.data
					};
				})
				.then((response) => {
					return dispatch(loadReposData(GET_REPOS_SUCCESS, response.data));
				})
				.catch((error) => dispatch(loadReposData(GET_REPOS_ERROR, error)));
		};
	}
};

export const selectors = {
	getReposData: (state) => state.repos.data,
	getReposError: (state) => state.repos.error
};
