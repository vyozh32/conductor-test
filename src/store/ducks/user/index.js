import axios from 'axios';
import endpoints from '../../../endpoints';
import { GET_REPOS_SUCCESS } from '../repos';

// Actions
const GET_USER_SUCCESS = 'conductor/test/get-user-data';
const GET_USER_ERROR = 'conductor/test/get-user-data-error';

// defaultState

const defaultUserState = {
	data: null,
	error: null
};

// Reducer
export default function(state = defaultUserState, action) {
	const { type, payload } = action;
	switch (type) {
		case GET_USER_SUCCESS: {
			return {
				...state,
				data: payload,
				error: null
			};
		}
		case GET_USER_ERROR: {
			return {
				...state,
				error: payload
			};
		}
		default: {
			return state;
		}
	}
}

// Action Creators

const loadUserData = (type, payload) => {
	return {
		type: type,
		payload
	};
};

// side effects, only as applicable
// e.g. thunks, epics, etc

export const actions = {
	getUserData: (user) => {
		return (dispatch, getState) => {
			const url = endpoints.USER.replace('{user_id}', user);
			axios
				.get(url)
				.then((res) => {
					return {
						data: res.data
					};
				})
				.then((response) => {
					return dispatch(loadUserData(GET_USER_SUCCESS, response.data));
				})
				.then((action) => dispatch({ type: GET_REPOS_SUCCESS, payload: null }))
				.catch((error) => {
					dispatch({ type: GET_USER_SUCCESS, payload: null });
					dispatch(loadUserData(GET_USER_ERROR, error));
				});
		};
	}
};

export const selectors = {
	getUserData: (state) => state.user.data,
	getUserError: (state) => state.user.error
};
