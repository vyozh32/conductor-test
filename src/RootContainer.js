import { Form, Alert, Row, Col, Grid, Glyphicon, Button } from 'react-bootstrap';
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import UserForm from './components/UserForm';
import UserProfile from './components/UserProfile';
import ReposList from './components/ReposList';
import { selectors as userSelectors, actions as userActions } from './store/ducks/user';
import { selectors as reposSelectors, actions as reposActions } from './store/ducks/repos';

import logo from './logo.svg';
import './Root.css';

class RootContainer extends Component {
	constructor(props) {
		super(props);
	}
	render() {
		const { userData, userError, getUserData, reposData, getReposData } = this.props;
		return (
			<div className="app">
				<header className="app-header">
					<img src={logo} className="app-logo" alt="logo" />
					<p>
						<code>Welcome to Github App</code>
					</p>
				</header>
				<div className="container container--form-wrapper">
					<div className="row">
						<div className="col-md-6 d-flex">
							<h3 className="text-left font-weight-bold">Insert user nickname</h3>
						</div>
					</div>
					<div className="row">
						<div className="col-md-12 mx-auto">
							<UserForm userError={userError} onChange={getUserData} />
						</div>
					</div>
					{userData ? <UserProfile {...userData} /> : null}
					<div className="row">
						{userData ? <ReposList reposData={reposData} getReposData={getReposData} /> : null}
					</div>
				</div>
			</div>
		);
	}
}

export default connect(
	(state) => ({
		userData: userSelectors.getUserData(state),
		userError: userSelectors.getUserError(state),
		reposData: reposSelectors.getReposData(state),
		reposError: reposSelectors.getReposError(state)
	}),
	(dispatch) => bindActionCreators({ ...userActions, ...reposActions }, dispatch)
)(RootContainer);
