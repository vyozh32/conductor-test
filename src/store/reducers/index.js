import {combineReducers} from 'redux'
import user from '../ducks/user'
import repos from '../ducks/repos';

export default combineReducers({
    user,
    repos
})