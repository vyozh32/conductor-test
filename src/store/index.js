// imports important items for store
import reducers from './reducers';
import thunk from 'redux-thunk';
import { createStore, applyMiddleware, compose } from 'redux';

const dev = process.env.NODE_ENV === 'development';
const devtools = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__;
const composeEnhancers = dev && devtools ? devtools : compose;

const middleware =  [thunk];

const createStoreWithMiddleware = composeEnhancers(applyMiddleware(
        ...middleware  // Saving done here
    ))(createStore)

const store = createStoreWithMiddleware(
	reducers,
);

export default store;