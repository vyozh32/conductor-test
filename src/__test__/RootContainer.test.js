import React from 'react';
import RootContainer from '../RootContainer';
import store from '../store';
import renderer from 'react-test-renderer';
import { GET_USER_SUCCESS } from '../store/ducks/user/index';
import { GET_REPOS_SUCCESS, actions as reposActions } from '../store/ducks/repos';

describe('<RootContainer>', function() {
	it('initialy renders', () => {
		const component = renderer.create(<RootContainer store={store} />);
		expect(component.toJSON()).toMatchSnapshot();
	});

	it('renders with user data', () => {
		const component = renderer.create(<RootContainer store={store} />);
		store.dispatch({
			type: GET_USER_SUCCESS,
			payload: {
				avatar_url: 'https://some-avatar-url.com',
				location: 'test location',
				followers: 12,
				following: 43,
				name: 'Test name',
				public_repos: 5,
				login: 'user_test'
			}
		});
		expect(component.toJSON()).toMatchSnapshot();
	});

	it('renders with repos data', () => {
		const component = renderer.create(<RootContainer store={store} />);
		store.dispatch({
			type: GET_REPOS_SUCCESS,
			payload: [
				{
					id: 1,
					name: 'Repos name',
					description: 'repos desc some test text',
					archived: 10,
					created_at: '322 228 1488 1337',
					fork: 10,
					forks: 100,
					name: 'Repos name',
					open_issues: 228,
					index: 10
				},
				{
					id: 2,
					name: 'Repos name',
					description: 'desc some test text',
					archived: 20,
					created_at: '322 228 1488 1337',
					fork: 6,
					forks: 23,
					name: 'Repos name',
					open_issues: 228,
					index: 32
				}
			]
		});
		expect(component.toJSON()).toMatchSnapshot();
	});
});
