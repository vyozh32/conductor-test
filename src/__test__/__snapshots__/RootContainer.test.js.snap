// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RootContainer> initialy renders 1`] = `
<div
  className="app"
>
  <header
    className="app-header"
  >
    <img
      alt="logo"
      className="app-logo"
      src="logo.svg"
    />
    <p>
      <code>
        Welcome to Github App
      </code>
    </p>
  </header>
  <div
    className="container container--form-wrapper"
  >
    <div
      className="row"
    >
      <div
        className="col-md-6 d-flex"
      >
        <h3
          className="text-left font-weight-bold"
        >
          Insert user nickname
        </h3>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-md-12 mx-auto"
      >
        <div
          className="form"
        >
          <div
            className="input-group mb-6"
          >
            <div
              className="input-group-prepend"
            >
              <span
                className="input-group-text"
                id="basic-addon1"
              >
                @
              </span>
            </div>
            <input
              aria-describedby="basic-addon1"
              aria-label="Username"
              className="form-control"
              defaultValue="vyozh32"
              placeholder="Username"
              type="text"
            />
            <button
              className="btn btn-primary"
              onClick={[Function]}
              type="submit"
            >
              Submit
            </button>
          </div>
          <div
            className="error-block"
          />
        </div>
      </div>
    </div>
    <div
      className="row"
    />
  </div>
</div>
`;

exports[`<RootContainer> renders with repos data 1`] = `
<div
  className="app"
>
  <header
    className="app-header"
  >
    <img
      alt="logo"
      className="app-logo"
      src="logo.svg"
    />
    <p>
      <code>
        Welcome to Github App
      </code>
    </p>
  </header>
  <div
    className="container container--form-wrapper"
  >
    <div
      className="row"
    >
      <div
        className="col-md-6 d-flex"
      >
        <h3
          className="text-left font-weight-bold"
        >
          Insert user nickname
        </h3>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-md-12 mx-auto"
      >
        <div
          className="form"
        >
          <div
            className="input-group mb-6"
          >
            <div
              className="input-group-prepend"
            >
              <span
                className="input-group-text"
                id="basic-addon1"
              >
                @
              </span>
            </div>
            <input
              aria-describedby="basic-addon1"
              aria-label="Username"
              className="form-control"
              defaultValue="vyozh32"
              placeholder="Username"
              type="text"
            />
            <button
              className="btn btn-primary"
              onClick={[Function]}
              type="submit"
            >
              Submit
            </button>
          </div>
          <div
            className="error-block"
          />
        </div>
      </div>
    </div>
    <div
      className="user-container"
    >
      <div
        className="row"
      >
        <div
          className="col-md-4"
        >
          <div
            className="user-profile d-flex"
          >
            <div
              className="avatar flex-column"
            >
              <img
                className="rounded"
                src="https://some-avatar-url.com"
              />
              <h3
                className="font-weight-bold"
              >
                Test name
              </h3>
              <h4
                className="p-2"
              >
                test location
              </h4>
              <p
                className="text-secondary"
              >
                user_test
              </p>
            </div>
          </div>
        </div>
        <div
          className="col-md-8 flex-column"
        >
          <div
            className="row"
          >
            <div
              className="col-md-4 p-2"
            >
              <span
                className="text-info font-weight-bold"
              >
                Followers 12
              </span>
            </div>
            <div
              className="col-md-4 p-2"
            >
              <span
                className="text-info font-weight-bold"
              >
                Following 43
              </span>
            </div>
            <div
              className="col-md-4 p-2"
            >
              <span
                className="text-info font-weight-bold"
              >
                Repositories 5
              </span>
            </div>
          </div>
          <div
            className="row"
          />
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-md-12 d-flex"
      >
        <button
          className="btn btn-big btn-primary"
          onClick={[Function]}
        >
          Get repositories
        </button>
      </div>
      <div
        className="repos-list col-md-12"
      >
        <table
          class="table table-striped table-dark table-hover"
        >
          <thead>
            <tr>
              <th
                scope="col"
              >
                #
              </th>
              <th
                scope="col"
              >
                Name
              </th>
              <th
                scope="col"
              >
                Description
              </th>
              <th
                scope="col"
              >
                Archived
              </th>
              <th
                scope="col"
              >
                Created_at
              </th>
              <th
                scope="col"
              >
                Fork
              </th>
              <th
                scope="col"
              >
                Forks
              </th>
              <th
                scope="col"
              >
                Issues
              </th>
            </tr>
          </thead>
          <tbody>
            <tr
              className="table-row"
              onClick={[Function]}
            >
              <td
                colspan="8"
                scope="col"
              >
                Repos name
              </td>
            </tr>
            <tr
              className="table-row"
              onClick={[Function]}
            >
              <td
                colspan="8"
                scope="col"
              >
                Repos name
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;

exports[`<RootContainer> renders with user data 1`] = `
<div
  className="app"
>
  <header
    className="app-header"
  >
    <img
      alt="logo"
      className="app-logo"
      src="logo.svg"
    />
    <p>
      <code>
        Welcome to Github App
      </code>
    </p>
  </header>
  <div
    className="container container--form-wrapper"
  >
    <div
      className="row"
    >
      <div
        className="col-md-6 d-flex"
      >
        <h3
          className="text-left font-weight-bold"
        >
          Insert user nickname
        </h3>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-md-12 mx-auto"
      >
        <div
          className="form"
        >
          <div
            className="input-group mb-6"
          >
            <div
              className="input-group-prepend"
            >
              <span
                className="input-group-text"
                id="basic-addon1"
              >
                @
              </span>
            </div>
            <input
              aria-describedby="basic-addon1"
              aria-label="Username"
              className="form-control"
              defaultValue="vyozh32"
              placeholder="Username"
              type="text"
            />
            <button
              className="btn btn-primary"
              onClick={[Function]}
              type="submit"
            >
              Submit
            </button>
          </div>
          <div
            className="error-block"
          />
        </div>
      </div>
    </div>
    <div
      className="user-container"
    >
      <div
        className="row"
      >
        <div
          className="col-md-4"
        >
          <div
            className="user-profile d-flex"
          >
            <div
              className="avatar flex-column"
            >
              <img
                className="rounded"
                src="https://some-avatar-url.com"
              />
              <h3
                className="font-weight-bold"
              >
                Test name
              </h3>
              <h4
                className="p-2"
              >
                test location
              </h4>
              <p
                className="text-secondary"
              >
                user_test
              </p>
            </div>
          </div>
        </div>
        <div
          className="col-md-8 flex-column"
        >
          <div
            className="row"
          >
            <div
              className="col-md-4 p-2"
            >
              <span
                className="text-info font-weight-bold"
              >
                Followers 12
              </span>
            </div>
            <div
              className="col-md-4 p-2"
            >
              <span
                className="text-info font-weight-bold"
              >
                Following 43
              </span>
            </div>
            <div
              className="col-md-4 p-2"
            >
              <span
                className="text-info font-weight-bold"
              >
                Repositories 5
              </span>
            </div>
          </div>
          <div
            className="row"
          />
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-md-12 d-flex"
      >
        <button
          className="btn btn-big btn-primary"
          onClick={[Function]}
        >
          Get repositories
        </button>
      </div>
      <div
        className="repos-list col-md-12"
      >
        <table
          class="table table-striped table-dark table-hover"
        >
          <thead>
            <tr>
              <th
                scope="col"
              >
                #
              </th>
              <th
                scope="col"
              >
                Name
              </th>
              <th
                scope="col"
              >
                Description
              </th>
              <th
                scope="col"
              >
                Archived
              </th>
              <th
                scope="col"
              >
                Created_at
              </th>
              <th
                scope="col"
              >
                Fork
              </th>
              <th
                scope="col"
              >
                Forks
              </th>
              <th
                scope="col"
              >
                Issues
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <th
                colspan="8"
                scope="row"
              >
                No data
              </th>
              <td />
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
